File: client-sdk/ts/electron-conversation-agent/src/sdk-usage.ts
import {
  createEvaVoiceDialogueAgent,
  type AgentEvent,
  type CommandRegistration,
  type EvaVoiceDialogueAgent,
} from "@autoark-ai/eva-client-sdk-ts";
import type { MediaTransportsConfig } from "@autoark-ai/eva-client-sdk-ts/spi";
import {
  createBrowserAudioInputSource,
  createBrowserAudioOutputSink,
  createBrowserCameraSnapshotSource,
  createPassthroughAecProcessor,
} from "@autoark-ai/eva-client-sdk-ts/browser";

export type { AgentEvent, EvaVoiceDialogueAgent };

type StartEvaAgentOptions = {
  apiKey: string;
  audioInputEnabled: boolean;
  ttsEnabled: boolean;
  cameraEnabled: boolean;
  onEvent: (event: AgentEvent, source: EvaVoiceDialogueAgent) => void;
};

type StartedEvaAgent = {
  agent: EvaVoiceDialogueAgent;
  unsubscribe: () => void;
};

type PageTheme = "light" | "dark";

const demoCommandRegistrations: readonly CommandRegistration[] = [
  {
    definition: {
      name: "show_current_time",
      description: "当用户询问当前时间时，查询当前页面所在设备的本地时间。",
    },
    handler(_call, context) {
      if (context.signal.aborted) {
        return { ok: false, message: "Command cancelled" };
      }
      const now = new Date();
      return {
        ok: true,
        message: `当前本地时间是 ${now.toLocaleTimeString()}`,
        data: {
          isoTime: now.toISOString(),
          localTime: now.toLocaleString(),
          timeZone: Intl.DateTimeFormat().resolvedOptions().timeZone,
        },
      };
    },
  },
  {
    definition: {
      name: "set_page_theme",
      description: "当用户要求切换页面外观时，将当前页面切换为指定的明暗主题。",
      parameters: [{
        name: "theme",
        description: "要应用的页面主题。",
        type: "string",
        required: true,
        enum: ["light", "dark"],
        example: "dark",
      }],
    },
    handler(call, context) {
      if (context.signal.aborted) {
        return { ok: false, message: "Command cancelled" };
      }
      const theme = call.arguments.theme;
      if (theme !== "light" && theme !== "dark") {
        return { ok: false, message: "Unsupported page theme" };
      }
      applyPageTheme(theme);
      return {
        ok: true,
        message: `页面主题已切换为 ${theme}`,
        data: { theme },
      };
    },
  },
];

function applyPageTheme(theme: PageTheme): void {
  document.documentElement.dataset.theme = theme;
  document.documentElement.style.colorScheme = theme;
}

/**
 * SDK 接入主路径（Electron 渲染进程语音 demo）。
 *
 * Electron 渲染进程运行在 Chromium 中，因此照常复用浏览器 Media SPI helper。
 * 页面弹窗、按钮状态、日志和消息渲染都留在 main.ts；这里连续展示：
 * 1. 组合 Media SPI；2. 创建 Agent；3. 订阅事件；4. 配置初始状态并启动。
 */
export async function startEvaAgent(options: StartEvaAgentOptions): Promise<StartedEvaAgent> {
  // 语音接入需要完整的 input / output / AEC；纯文本接入可以省略 transports 与 vad。
  const transports: MediaTransportsConfig = {
    input: createBrowserAudioInputSource({
      maxBufferedChunks: 64,
      echoCancellation: true,
      noiseSuppression: true,
      autoGainControl: true,
    }),
    output: createBrowserAudioOutputSink(),
    aec: createPassthroughAecProcessor(),
    // 默认 browser camera helper 是可替换的 CameraSnapshotSource；factory 不会自行申请权限。
    camera: createBrowserCameraSnapshotSource({
      video: {
        facingMode: { ideal: "user" },
        width: { ideal: 640 },
        height: { ideal: 360 },
      },
      mimeType: "image/jpeg",
      jpegQuality: 0.7,
    }),
  };

  // apiKey 与三个 model 是应用提供的接入配置；其余字段是本 demo 的选择。
  const agent = createEvaVoiceDialogueAgent({
    apiKey: options.apiKey,
    asr: { model: "ark-asr-plus", sampleRate: 16_000 },
    llm: {
      model: "volcengine-doubao-seed-2.0-lite",
      extraParameters: {
        thinking: {
          type: "disabled",
        },
      },
    },
    tts: {
      model: "ark-tts-flash",
      voice: "zh_en_male_evan",
      sampleRate: 44_100,
    },
    vad: {
      sensitivity: 0.7,
      silenceThresholdMs: 400,
    },
    transports,
    // 与 browser demo 保持一致：首段播放期间给 AEC 留出稳定时间，避免被自己的开场音频打断。
    bargeIn: {
      initialPlaybackGuardMs: 3000,
    },
    history: { maxTurns: 10 },
    camera: { captureTimeoutMs: 1500 },
    // 可选：省略 emotion（或设为 enabled: false）时不运行旁路分类，也不产生 emotion.detected。
    // custom labels 会替换默认业务标签；SDK 会自动补充 unknown。
    emotion: { enabled: true, labels: ["happy", "sad"] },
    // 可选：省略 commands（或提供空 registrations）时不向 LLM 暴露 command。
    commands: {
      registrations: demoCommandRegistrations,
      maxCallsPerTurn: 3,
    },
    systemPrompt: [
      "你是 EVA，一个简洁、自然、可靠的语音与文字对话助手。",
      "用户的语音会先由系统转写成当前消息再交给你；不要说自己听不到用户，也不要否认正在进行语音对话。",
      "当用户询问你是否听到时，说明你能处理已经送达的语音转写；不要声称能直接感知尚未发送的声音。",
      "只有当前用户消息实际包含图片时，才可以描述当前看到的画面。",
      "摄像头开启且当前用户消息带有图片时，这张图片就是你此刻看到的画面；应基于它回答，不要把它当成历史图片，也不要否认你收到了视觉输入。",
      "当用户询问你能看到什么但当前消息没有图片时，明确说明当前没有收到新的图片或摄像头画面；不要声称看到了现实画面，也不要笼统说自己完全不能看图。",
      "当前用户消息没有图片时，不得把历史中的视觉描述说成实时观察；",
      "如需引用，只能明确说明那是之前看到的内容，并说明当前没有新的画面。",
    ].join(""),
    greeting: {
      mode: "static",
      text: [
        "你好，很高兴见到你！我是 EVA 语音助手，接下来你可以和我随便聊聊天，",
        "也可以问我现在的时间、让我切换页面主题，或者打开摄像头后问我看到了什么。",
        "我会认真听你说话，并尽量用简洁自然的方式回答。",
        "准备好以后，直接对我说话就可以了。",
      ].join(""),
    },
  });

  const unsubscribe = agent.onEvent((event) => options.onEvent(event, agent));
  try {
    await agent.setAudioInputEnabled(options.audioInputEnabled);
    await agent.setCameraCaptureEnabled(options.cameraEnabled);
    await agent.setTtsEnabled(options.ttsEnabled);
    await agent.start();
    return { agent, unsubscribe };
  } catch (error) {
    unsubscribe();
    throw error;
  }
}

File: client-sdk/flutter/conversation-agent/lib/sdk_usage.dart
import 'dart:async';

import 'package:autoark_eva_client_sdk/autoark_eva_client_sdk.dart';

const String demoSystemPrompt = '''
你是 EVA，一个正在通过实时语音与用户交流的智能助手。
用户消息可能由语音识别转写而来，可能包含口语、省略、同音字或少量识别错误。请结合上下文理解用户意图，优先用自然、简洁、适合直接朗读的口语回答；不要把交互描述成打字、文本聊天或“阅读消息”，除非用户明确谈论文字。
当本轮消息附带图片时，该图片来自用户当前设备的摄像头，代表用户此刻让你看到的画面。用户问“你看到了什么”“这是什么”等问题时，应直接根据图片内容回答，就像你正通过摄像头观察现场；不要要求用户再次上传图片，也不要声称无法访问已经附带的摄像头画面。
如果语音转写或画面信息不足以可靠判断，请自然地向用户追问，不要编造。
''';

final class DemoConfiguration {
  const DemoConfiguration({
    required this.apiKey,
    this.asrModel = 'ark-asr-plus',
    this.llmModel = 'volcengine-doubao-seed-2.0-mini',
    this.ttsModel = 'ark-tts-flash',
  });

  factory DemoConfiguration.fromEnvironment() => const DemoConfiguration(
    apiKey: String.fromEnvironment('EVA_GATEWAY_API_KEY'),
    llmModel: String.fromEnvironment(
      'EVA_GATEWAY_LLM_MODEL',
      defaultValue: 'volcengine-doubao-seed-2.0-mini',
    ),
    ttsModel: String.fromEnvironment(
      'EVA_GATEWAY_TTS_MODEL',
      defaultValue: 'ark-tts-flash',
    ),
  );

  final String apiKey;
  final String asrModel;
  final String llmModel;
  final String ttsModel;

  DemoConfiguration withApiKey(String value) => DemoConfiguration(
    apiKey: value,
    asrModel: asrModel,
    llmModel: llmModel,
    ttsModel: ttsModel,
  );

  bool get isReady =>
      apiKey.trim().isNotEmpty &&
      asrModel.trim().isNotEmpty &&
      llmModel.trim().isNotEmpty &&
      ttsModel.trim().isNotEmpty;
}

/// SDK 接入主路径：把应用配置转换为一个使用 Android 默认媒体 helper 的 EVA Agent。
DemoAgentPort createEvaDemoAgent(DemoConfiguration config) {
  final EvaAgent agent = EvaAgent.create(buildEvaAgentConfig(config));
  return EvaDemoAgentPort._(agent);
}

/// Demo 选择的完整 public SDK 配置；UI 不参与配置拼装。
EvaAgentConfig buildEvaAgentConfig(DemoConfiguration config) => EvaAgentConfig(
  // 必填：Gateway AK 与 ASR、LLM、TTS model 由接入应用提供。
  apiKey: config.apiKey,
  asr: EvaAsrConfig(model: config.asrModel, sampleRate: 16000),
  llm: EvaLlmConfig(
    model: config.llmModel,
    extraParameters: const <String, Object?>{
      'thinking': <String, Object?>{'type': 'disabled'},
    },
  ),
  tts: EvaTtsConfig(
    model: config.ttsModel,
    voice: 'zh_en_male_evan',
    sampleRate: 44100,
  ),

  // 省略 transports，Android 会使用 SDK 自带的麦克风、扬声器、AEC 与 CameraX helper。
  vad: const EvaVadConfig(sensitivity: 0.7, silenceThresholdMs: 400),
  history: const EvaHistoryConfig(maxTurns: 10),
  camera: const EvaCameraConfig(
    captureTimeoutMs: 1500,
    maxLongEdge: 640,
    jpegQuality: 70,
  ),
  bargeIn: const EvaBargeInConfig(initialPlaybackGuardMs: 3000),
  systemPrompt: demoSystemPrompt,
  greeting: const EvaStaticGreeting('你好，我是 EVA，很高兴认识你。'),
  metadata: const <String, Object?>{'surface': 'flutter-demo'},

  // 可选：省略 emotion 时不会运行旁路分类，也不会产生 emotion.detected。
  emotion: EvaEmotionConfig(enabled: true),

  // 可选：省略 commands 时不会向 LLM 暴露 command definitions 或执行 handler。
  commands: EvaCommandsConfig(
    maxCallsPerTurn: 3,
    registrations: _buildDemoCommands(),
  ),
);

List<EvaCommandRegistration> _buildDemoCommands() => <EvaCommandRegistration>[
  EvaCommandRegistration(
    definition: EvaCommandDefinition(
      name: 'get_current_time',
      description: '获取设备当前时间。用户询问当前时间时调用。',
    ),
    handler: (_, _) {
      final DateTime now = DateTime.now();
      return EvaCommandSuccess(
        message: '设备当前时间为 ${now.toIso8601String()}。',
        data: <String, Object?>{'iso': now.toIso8601String()},
      );
    },
  ),
  EvaCommandRegistration(
    definition: EvaCommandDefinition(
      name: 'format_greeting',
      description: '根据称呼与语气生成一条问候语，不执行外部业务副作用。',
      parameters: <EvaCommandParameter>[
        EvaStringCommandParameter(
          name: 'tone',
          description: '问候语气。',
          required: true,
          enumValues: const <String>['formal', 'casual'],
        ),
        EvaStringCommandParameter(
          name: 'name',
          description: '问候对象。',
          required: true,
        ),
      ],
    ),
    handler: (EvaCommandCall call, _) {
      final String name = call.arguments['name'] as String;
      final String tone = call.arguments['tone'] as String;
      return EvaCommandSuccess(
        message: tone == 'formal' ? '尊敬的 $name，您好。' : '$name，你好！',
      );
    },
  ),
];

abstract interface class DemoAgentPort {
  Stream<EvaAgentEvent> get events;

  Future<void> start();
  Future<void> stop();
  Future<void> submitText(String text);
  Future<void> setAudioInputEnabled(bool enabled);
  Future<void> setCameraEnabled(bool enabled);
  Future<void> setTtsEnabled(bool enabled);
  Future<List<EvaConversationMessage>> getMessages();
}

typedef DemoAgentFactory = DemoAgentPort Function(DemoConfiguration config);

final class EvaDemoAgentPort implements DemoAgentPort {
  EvaDemoAgentPort._(this._agent);

  final EvaAgent _agent;

  @override
  Stream<EvaAgentEvent> get events => _agent.events;

  @override
  Future<List<EvaConversationMessage>> getMessages() => _agent.getMessages();

  @override
  Future<void> setAudioInputEnabled(bool enabled) =>
      _agent.setAudioInputEnabled(enabled);

  @override
  Future<void> setCameraEnabled(bool enabled) =>
      _agent.setCameraEnabled(enabled);

  @override
  Future<void> setTtsEnabled(bool enabled) => _agent.setTtsEnabled(enabled);

  @override
  Future<void> start() => _agent.start();

  @override
  Future<void> stop() => _agent.stop();

  @override
  Future<void> submitText(String text) => _agent.submitText(text);
}
